artiq/coredevice/mirny.py

```diff
@@ -140,3 +151,232 @@ def write_ext(self, addr, length, data):
+        self._send_mezz_data([
+            (ALMAZNY_REG_CLEAR, 1),
+            (ALMAZNY_REG_LATCH_BASE + ch, 0)
+        ])
+
+    @kernel
+    def reg_clear_all(self):
+        """
+        Clears all registers.
+        """
+        self._send_mezz_data([
+            # everything 0 including REG_CLEAR
+        ])
+        self._send_mezz_data([
+            (ALMAZNY_REG_CLEAR, 0),
+            (ALMAZNY_REG_LATCH_BASE, 1),
+            (ALMAZNY_REG_LATCH_BASE + 1, 1),
+            (ALMAZNY_REG_LATCH_BASE + 2, 1),
+            (ALMAZNY_REG_LATCH_BASE + 3, 1),
+        ])
+        self._send_mezz_data([
+            (ALMAZNY_REG_CLEAR, 1),
+        ])
+
+    @kernel
+    def _cycle(self, data):
+        """
+        one cycle for inputting register data
+        """
+        self._send_mezz_data([
+            (ALMAZNY_REG_CLEAR, 1),
+            # clk = 0
+        ] + data)
+        self._send_mezz_data([
+            (ALMAZNY_REG_CLEAR, 1),
+            (ALMAZNY_SER_CLK, 1),
+        ] + data)
+
+    @kernel
+    def _latch(self, ch):
+        self._send_mezz_data([
+            (ALMAZNY_REG_CLEAR, 1),
+            (ALMAZNY_SER_CLK, 1),
+            (ALMAZNY_REG_LATCH_BASE + ch, 1)
+        ])
+        self._send_mezz_data([
+            (ALMAZNY_REG_CLEAR, 1),
+            # reset latch
+        ])
+
+    @kernel
+    def _latch_all(self):
+        self._send_mezz_data([
+            (ALMAZNY_REG_CLEAR, 1),
+            (ALMAZNY_SER_CLK, 1),
+            (ALMAZNY_REG_LATCH_BASE + 0, 1),
+            (ALMAZNY_REG_LATCH_BASE + 1, 1),
+            (ALMAZNY_REG_LATCH_BASE + 2, 1),
+            (ALMAZNY_REG_LATCH_BASE + 3, 1)
+        ])
+        self._send_mezz_data([
+            (ALMAZNY_REG_CLEAR, 1),
+            # reset all latches
+        ])
+
+    @kernel
+    def _update_register(self, ch):
+        data = (self.rf_switches[ch] << 6) | self.att_mu[ch]
+
+        self._send_mezz_data([
+            (ALMAZNY_REG_CLEAR, 1),
+            # reset all latches
+        ])
+
+        for i in range(8):
+            self._cycle([
+                (ALMAZNY_SER_MOSI, (data >> (7 - i)) & 0x01) 
+            ])
+        self._latch(ch)
+
+    @kernel
+    def _update_all_registers(self):
+        # assumes data is identical for all channesls
+        data = (self.rf_switches[0] << 6) | self.att_mu[0]
+
+        self._send_mezz_data([
+            (ALMAZNY_REG_CLEAR, 1),
+            # reset all latches
+        ])
+        for i in range(8):
+            self._cycle([
+                (ALMAZNY_SER_MOSI, (data >> (7 - i)) & 0x01) 
+            ])
+        self._latch_all()
+
+    @kernel
+    def _send_mezz_data(self, pins_data):
+        """
+        Sends the raw data to the mezzanine board.
+        :param pins_data - list of tuples in format (pin_number, bit)
+        """
+        # by default data is 0 - put 1s where you need
+        data = ALMAZNY_OE_MASK
+        for pin, d in pins_data:
+            data = (data & ~(1 << pin)) | d << pin
```

Comment:
Do we really need this list business? Bitwise operations aren't much more complicated to write than adding list elements and it would make the code a bit simpler and consistent with the way most low-level code is written. The runtime speed could be improved too.